fix(gux-form-field): improve label accessibility

katie-bobbe-genesys opened the pull request.

katie-bobbe-genesys commented:
Not ready for review yet. I just wanted to create a feature branch so that I can test my work in progress on my virtual box with NVDA and JAWS.

daragh-king-genesys commented:
> Not ready for review yet. I just wanted to create a feature branch so that I can test my work in progress on my virtual box with NVDA and JAWS.

You don't need to create a PR to get a time machine branch. They are created in Jenkins by a job linked to new branches and branch updates.

katie-bobbe-genesys commented on src/components/stable/gux-form-field/gux-form-field.tsx:
```diff
@@ -360,14 +384,14 @@ export class GuxFormField {
   }
 
   private getError(hasError: boolean): JSX.Element {
```
Is there any reason why we would need to create a div with a class of `gux-error` and a randomly generated id? This causes there to be an empty div created in each component, and the input is provided with an `aria-describedby` attribute that points to this empty div. Let me know if I'm missing something by refactoring this.

katie-bobbe-genesys commented:
I am still working on tests but I am opening up this PR draft for review. Some of the main points that were covered by this PR include

**Label Accessibility**: It is no longer required to provide a `for` attribute on the label or an `id` on the input. The component will generate a random `id` that will link the label and input for accessibility.

- If a user does not provide an `id` or `for` attribute, a randomly generated `id` is provided to the input, and a `for` attribute linking to that `id` is added to the label
- If a user does not provide a `for` attribute on the label, but there is an `id` on the input, a `for` attribute linking to that existing `id` will be added to the label
- If a user does not provide an `id` on the input but does provide a `for` attribute on the label, a console error will show asking the user to either provide an `id` on the input to match the `for` attribute or omit the `for` attribute to allow the component to automatically generate the `id` and `for` attribute
- If a user provides an `id` on the input and a corresponding `for` attribute on the label, no action is taken

**Label Position**: Users can now specify the position of the label using the `label-position` attribute (above or beside)

- Previously, a label was shown beside the input when the `offsetWidth` of the label was less than 40, and shown above the input if longer
- If no `label-position` attribute is provided, the label position will function the same as previously (no breaking changes)

**Linking the error message with the input**: The error message is conveyed to the screenreader using `aria-describedby`

- Fixed a bug where the error message content was not being conveyed to the screenreader. As a side note, I have found that `aria-describedby` does not get read by Voiceover, but is read by NVDA and JAWS.

**Warning when no label slot is provided**: An error will be logged to the console if a label is not provided to a form field component

**Updated examples**: Updated the examples in the docs

-  Removed `ids` and `for` attributes on the inputs and labels
- Provided `label-position` attributes
- Clarified a few inconsistencies

daragh-king-genesys commented:
> * If a user does not provide an id on the input but does provide a for attribute on the label, a console error will show asking the user to either provide an id on the input to match the for attribute or omit the for attribute to allow the component to automatically generate the id and for attribute

Should we also log an error if they provide an id and for values that do not match?

daragh-king-genesys commented on src/components/stable/gux-form-field/gux-form-field.tsx:
```diff
@@ -17,7 +17,10 @@ export class GuxFormField {
   private requiredObserver: MutationObserver;
-  private errorID = randomHTMLId('gux-form-field');
+  private errorId = randomHTMLId('gux-form-field');
+  private labelId = randomHTMLId('gux-form-field');
```
Should this be called `defaultLabelId`? and prefixed

daragh-king-genesys commented on src/components/stable/gux-form-field/gux-form-field.tsx:
```diff
@@ -17,7 +17,10 @@ export class GuxFormField {
   private label: HTMLLabelElement;
   private requiredObserver: MutationObserver;
-  private errorID = randomHTMLId('gux-form-field');
+  private errorId = randomHTMLId('gux-form-field');
```
I would not have the same prefix for the error and the label. I would use 'gux-form-field-error' here

daragh-king-genesys commented on src/components/stable/gux-form-field/gux-form-field.tsx:
```diff
@@ -72,14 +78,41 @@ export class GuxFormField {
+            ? 'beside'
+            : 'above';
+      }
+    }
```
Should we have an no label warning/error?

daragh-king-genesys commented on src/components/stable/gux-form-field/gux-form-field.tsx:
```diff
@@ -145,166 +174,161 @@ export class GuxFormField {
+          }}
+        >
+          <slot name="label" slot="label" />
+        </div>
```
Can 
```
<div
  class={{
    'gux-label-container': true,
    'gux-required': this.required
  }}
>
  <slot name="label" slot="label" />
</div>
```
 be centralised into a function?

daragh-king-genesys commented on src/components/stable/gux-form-field/gux-form-field.tsx:
```diff
@@ -72,14 +78,41 @@ export class GuxFormField {
+          this.label.offsetWidth > 1 && this.label.offsetWidth < 40
+            ? 'beside'
+            : 'above';
+      }
```
I would make this its own function called something like `setComputedLabelPosition`

daragh-king-genesys commented on src/components/stable/gux-form-field/gux-form-field.tsx:
```diff
@@ -72,14 +78,41 @@ export class GuxFormField {
+      );
+    }
+
+    this.inputHasId = !!this.input.hasAttribute('id');
```
I don't think `labelHasForAttribute` and `inputHasId` need to be instance variables. I think we can create a function that takes the input and label elements as inputs and does all the validation and fixing we need.

daragh-king-genesys commented on src/components/stable/gux-form-field/gux-form-field.tsx:
```diff
@@ -34,11 +37,14 @@ export class GuxFormField {
   valueInTooltip: boolean;
 
+  @Prop()
+  labelPosition: 'above' | 'beside';
```
I think we should add this to the component tracking.

daragh-king-genesys commented on src/components/stable/gux-form-field/gux-form-field.tsx:
```diff
@@ -72,14 +78,41 @@ export class GuxFormField {
+            : 'above';
+      }
+    }
+    if (!this.inputHasId && this.labelHasForAttribute) {
```
As I commented earlier I think this could be achieved with a function that takes in the input and label element.

daragh-king-genesys commented on src/components/stable/gux-form-field/gux-form-field.tsx:
```diff
@@ -72,14 +78,41 @@ export class GuxFormField {
+      const forId = this.input.getAttribute('id');
+      this.label.setAttribute('for', forId);
+    }
+    if (this.hasErrorSlot()) {
```
Does this need an else case to remove the attribute if not needed?

daragh-king-genesys replied:
Also should this be in a different lifecycle hook as it `hasErrorSlot` is changeable over time?

daragh-king-genesys reviewed (commented):
Looks good

katie-bobbe-genesys commented:
> Regarding label accessibility:
> 
> * what happens if someone puts a `span` in the label slot?
> * what happens if no label is provided?
> * did we also handle the `label-screenreader-only` case discussed in the comments of COMU-499?

- If someone puts a `span` in the label slot there seems to be no negative effects on appearance or accessibility. 
- If no label is provided, an error is logged to the console instructing the user to provide a label. 
- I just revisited the COMUI-499 and I now see the suggestion to add `screenreader-only` to the label position options, and I will add this now.

AThiyagarajan commented on src/components/stable/gux-form-field/gux-form-field.tsx:
```diff
@@ -72,14 +78,41 @@ export class GuxFormField {
+      throw new Error(
+        '[gux-form-field] A label is required for this component. If a visual label is not needed for this use case, please add localized text for a screenreader and use the gux-sr-only class to visually hide the label.'
+      );
+    }
```
I think throwing an error here will break executing the following codes right?

We had similar issues in the past for `gux-icons`, and the error was propagated to the host app.

So we had to throw the error on a `setTimeout` to overcome that.

```
    setTimeout(() => {
        throw err;
      }, 0);
```

I think our intention is to show an error on the browser console right?

katie-bobbe-genesys replied:
Thank you for pointing this out, I will look into this now. Yes, the intention is to show an error on the browser console.

daragh-king-genesys replied:
Also look into creating a utility function to standardise these error messages. Something like 
`logError(component: string, message: string): void {}`

MattCheely replied:
+1 on Daragh's comment about a utility function for that deferred throw pattern if we don't have one already.

daragh-king-genesys commented on src/components/stable/gux-form-field/gux-form-field.tsx:
```diff
@@ -78,42 +85,14 @@ export class GuxFormField {
-    if (this.hasErrorSlot()) {
-      this.input.setAttribute('aria-describedBy', this.errorId);
-    }
+    this.computedLabelPosition = this.setComputedLabelPosition(
```
function name and functionality seem a bit off to me.

If the function is called `setComputedLabelPosition` I would assume the assignment to `this.computedLabelPosition` happens in the function and all you would see at the call site is

```
this.computedLabelPosition = this.getComputedLabelPosition(this.label, this.labelPosition);
 ```

 For the functionality you have at the moment I would expect the function to be called `getComputedLabelPosition`

```
this.setComputedLabelPosition(this.label, this.labelPosition);
 ```

daragh-king-genesys commented on src/components/stable/gux-form-field/gux-form-field.tsx:
```diff
@@ -146,14 +125,7 @@ export class GuxFormField {
-        >
-          <slot name="label" slot="label" />
-        </div>
+        {this.getLabel()}
```
I would rename this `renderLabel`

Going forward I think all functions that return JSX should be called `render....`

daragh-king-genesys commented on src/components/stable/gux-form-field/gux-form-field.tsx:
```diff
@@ -379,10 +309,79 @@ export class GuxFormField {
     return !!this.root.querySelector('[slot="error"]');
   }
 
+  private getLabel(): JSX.Element {
```
Not sure why but my instinct here would be to pass in `this.required` as a parameter instead of using the instance variable in the function. Don't have a good reason why at the moment but it would make pulling all these `helper` functions out to a separate file in the future easier

daragh-king-genesys reviewed (approved):
👍
